### PR TITLE
Cm 1134 immediate contribution

### LINKED ISSUE #1134: Implement immediate contribution to common endpoint
**General idea**

Users should be able to contribute to commons freely whenever they want.

*User must be a common member to contribute to it

It should be stated in the request if it is a one time or recurring payment.

payment should function as all other payments in the backend (balance, payments collection, etc.)

if the user has already a card, use that one instead of sending a link to PayMe iframe.

*PayMe has an option to immediately charge and generate a buyer token that returns in the response, it might be more useful to use that ability in case the user doesn't have a card.

https://www1.payme.co.il/system/documentation#/introduction/api-capabilities-and-structure

https://app.zeplin.io/project/603377c3d4da061120ed2961/screen/61718640fdc4cda0f5be1797
https://app.zeplin.io/project/603377c3d4da061120ed2961/screen/61718641dc456ea4c8a51e77

### PR DESCRIPTION
**Pull request by ariel-dao**
**How to use:**
(example live in postman too)
POST to `{{url}}/commons/immediate-contribution`

body example:
```
{
    "amount": 1000,
    
    "commonId": "e8652509-f986-4dd0-81df-d31043727451",

    "contributionType": "monthly",

    "saveCard": true
}
```

User must be a member of the common (and be authenticated obviously)

Response can be either a payment entity (if the user has a card in the system, that card will be immediately used),
or a link with a payment Id, the link will be the iframe that the user will need to enter credit card details too, and the payment Id is the pending payment (front should listen to that payment Id to receive payment status)

* Monthly payments must have "saveCard" as true
* contributionType = "monthly" OR "one-time"

**Use cases and tests that were made in dev:**
contribute without existing card to common (one time) - save card - tested on dev
contribute without existing card to common (one time) - don’t save card - tested on dev
contribute with existing card to common (one time) - tested on dev
contribute with existing card to common (subscription) - tested on dev
contribute without existing card to common (subscription) - save card (default behavior) - tested on dev
contribute without existing card to common (subscription) - don’t save card (check for error)  - tested on dev

### CODE REVIEW

File: functions/src/common/business/immediateContribution.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                userId
+            })
+
+            await updateCommonBalance(commonId, Number(payment.amount.amount));
+        }
+    }
+
+
+    if (!card) {
+        if (!saveCard && contributionType === CONTRIBUTION.MONTHLY) throw new ArgumentError("cant create a subscription without saving card details");
```
Comment (noamso1) [inline functions/src/common/business/immediateContribution.ts:72]: you say "cant create a sub" but it was already created 20 lines ago

File: functions/src/payments/business/payme/payIn/createPaymentPage.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    userId,
+}: {
+    amount: number
+    paymentId: string
+    saveCard: boolean
+    contributionType: CONTRIBUTION
+    commonId: string
+    userId: string
+}) => {
+
```
Comment (noamso1) [inline functions/src/payments/business/payme/payIn/createPaymentPage.ts:29]: i personally like the straightforwardness here, but usually we store type in a separate folder...

File: functions/src/payments/controllers/payme/payIn/updatePaymentStatus.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import { responseExecutor } from '../../../../util';
+import { UnauthorizedError, NotFoundError, CommonError } from "../../../../util/errors";
+import { paymePayInService } from '../../../services/payme/payIn'
+import { md5Verification } from '../../../providers/payme/utils';
+import { IPaymeCallbackUpdate } from "../../../../shared/interfaces/payments/payme/IPaymeCallbackUpdate";
+import { redis } from "../../../../util/redis";
+import { PAYME_STATUS } from '../../../../shared/enums/payme/STATUSES';
+import { commonDb } from '../../../../common/database';
+import { CONTRIBUTION } from "../../../../shared/enums/CONTRIBUTION";
+
```
Comment (noamso1) [inline functions/src/payments/controllers/payme/payIn/updatePaymentStatus.ts:11]: please add a general description for all new files - file name "update payment status" is not enough.. who updates? when? etc..

Comment (noamso1) [reply functions/src/payments/controllers/payme/payIn/updatePaymentStatus.ts:11]: even this one line is good enough:
// gets an update back from Payme and updates the DB

File: functions/src/payments/services/payme/payIn/updatePaymentStatus.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            userId,
+            paymentMethod: 'card',
+            ...(
+                status_error_code && {
+                    error: {
+                        code: Number(status_error_code),
+                        details: status_error_details
+                    }
+                }
+            ),
```
Comment (noamso1) [inline functions/src/payments/services/payme/payIn/updatePaymentStatus.ts:94]: these last 20 lines ( whenever we paymentDB.add ) is repeated a few times very similarly. maybe we should create a reusable function for it ?

**Review by noamso1 — APPROVED** [review]

File: functions/src/payments/business/payme/payIn/createPaymentPage.ts (hunk 2 above)
Comment (ariel-dao) [reply functions/src/payments/business/payme/payIn/createPaymentPage.ts:29]: sometimes less is more

File: functions/src/payments/services/payme/payIn/updatePaymentStatus.ts (hunk 4 above)
Comment (ariel-dao) [reply functions/src/payments/services/payme/payIn/updatePaymentStatus.ts:94]: it's possible although it'll probably be the same length

File: functions/src/common/business/immediateContribution.ts (hunk 1 above)
Comment (ariel-dao) [reply functions/src/common/business/immediateContribution.ts:72]: that's a different conditional (assuming a card already exists)

File: functions/src/payments/controllers/payme/payIn/updatePaymentStatus.ts (hunk 3 above)
Comment (ariel-dao) [reply functions/src/payments/controllers/payme/payIn/updatePaymentStatus.ts:11]: added

**Pull request merged**